test: flaky TestRUMRoutingIntegration

https://github.com/elastic/apm-server/actions/runs/27533386489/job/81376673970

```
=== RUN   TestRUMRoutingIntegration
    rum_test.go:224:   []any{
        - 	map[string]any{
        - 		"@timestamp":                []any{string("dynamic")},
        - 		"agent.name":                []any{string("js-base")},
        - 		"agent.name.text":           []any{string("js-base")},
        - 		"agent.version":             []any{string("4.8.1")},
        - 		"client.ip":                 []any{string("dynamic")},
        - 		"data_stream.dataset":       []any{string("apm.rum")},
        - 		"data_stream.namespace":     []any{string("default")},
        - 		"data_stream.type":          []any{string("traces")},
        - 		"destination.address":       []any{string("localhost")},
        - 		"destination.port":          []any{float64(8003)},
        - 		"event.outcome":             []any{string("success")},
        - 		"event.success_count":       []any{float64(1)},
        - 		"http.request.method":       []any{string("POST")},
        - 		"http.response.status_code": []any{float64(200)},
        - 		"labels.testTagKey":         []any{string("testTagValue")},
        - 		"network.connection.type":   []any{string("5G")},
        - 		...
        - 	},
        - 	map[string]any{
        - 		"@timestamp":                []any{string("dynamic")},
        - 		"agent.name":                []any{string("js-base")},
        - 		"agent.name.text":           []any{string("js-base")},
        - 		"agent.version":             []any{string("4.8.1")},
        - 		"client.ip":                 []any{string("dynamic")},
        - 		"data_stream.dataset":       []any{string("apm.rum")},
        - 		"data_stream.namespace":     []any{string("default")},
        - 		"data_stream.type":          []any{string("traces")},
        - 		"destination.address":       []any{string("localhost")},
        - 		"destination.port":          []any{float64(8000)},
        - 		"event.outcome":             []any{string("success")},
        - 		"event.success_count":       []any{float64(1)},
        - 		"http.request.method":       []any{string("GET")},
        - 		"http.response.status_code": []any{float64(200)},
        - 		"labels.testTagKey":         []any{string("testTagValue")},
        - 		"network.connection.type":   []any{string("5G")},
        - 		...
        - 	},
        - 	map[string]any{
        - 		"@timestamp":              []any{string("dynamic")},
        - 		"agent.name":              []any{string("js-base")},
        - 		"agent.name.text":         []any{string("js-base")},
        - 		"agent.version":           []any{string("4.8.1")},
        - 		"client.ip":               []any{string("dynamic")},
        - 		"data_stream.dataset":     []any{string("apm.rum")},
        - 		"data_stream.namespace":   []any{string("default")},
        - 		"data_stream.type":        []any{string("traces")},
        - 		"event.outcome":           []any{string("unknown")},
        - 		"labels.testTagKey":       []any{string("testTagValue")},
        - 		"network.connection.type": []any{string("5G")},
        - 		"observer.hostname":       []any{string("dynamic")},
        - 		"observer.type":           []any{string("apm-server")},
        - 		"observer.version":        []any{string("dynamic")},
        - 		"parent.id":               []any{string("ec2e280be8345240")},
        - 		"processor.event":         []any{string("span")},
        - 		...
        - 	},
        - 	map[string]any{
        - 		"@timestamp":                []any{string("dynamic")},
        - 		"agent.name":                []any{string("js-base")},
        - 		"agent.name.text":           []any{string("js-base")},
        - 		"agent.version":             []any{string("4.8.1")},
        - 		"client.ip":                 []any{string("dynamic")},
        - 		"data_stream.dataset":       []any{string("apm.rum")},
        - 		"data_stream.namespace":     []any{string("default")},
        - 		"data_stream.type":          []any{string("traces")},
        - 		"destination.address":       []any{string("localhost")},
        - 		"destination.port":          []any{float64(8003)},
        - 		"event.outcome":             []any{string("success")},
        - 		"event.success_count":       []any{float64(1)},
        - 		"http.request.method":       []any{string("POST")},
        - 		"http.response.status_code": []any{float64(200)},
        - 		"labels.testTagKey":         []any{string("testTagValue")},
        - 		"network.connection.type":   []any{string("5G")},
        - 		...
        - 	},
          	map[string]any{"@timestamp": []any{string("dynamic")}, "agent.name": []any{string("js-base")}, "agent.name.text": []any{string("js-base")}, "agent.version": []any{string("4.8.1")}, ...},
        - 	map[string]any{
        - 		"@timestamp":              []any{string("dynamic")},
        - 		"agent.name":              []any{string("js-base")},
        - 		"agent.name.text":         []any{string("js-base")},
        - 		"agent.version":           []any{string("4.8.1")},
        - 		"client.ip":               []any{string("dynamic")},
        - 		"data_stream.dataset":     []any{string("apm.rum")},
        - 		"data_stream.namespace":   []any{string("default")},
        - 		"data_stream.type":        []any{string("traces")},
        - 		"event.outcome":           []any{string("success")},
        - 		"event.success_count":     []any{float64(1)},
        - 		"labels.testTagKey":       []any{string("testTagValue")},
        - 		"network.connection.type": []any{string("5G")},
        - 		"observer.hostname":       []any{string("dynamic")},
        - 		"observer.type":           []any{string("apm-server")},
        - 		"observer.version":        []any{string("dynamic")},
        - 		"parent.id":               []any{string("ec2e280be8345240")},
        - 		...
        - 	},
        - 	map[string]any{
        - 		"@timestamp":                      []any{string("dynamic")},
        - 		"agent.name":                      []any{string("js-base")},
        - 		"agent.name.text":                 []any{string("js-base")},
        - 		"agent.version":                   []any{string("4.8.1")},
        - 		"client.ip":                       []any{string("dynamic")},
        - 		"data_stream.dataset":             []any{string("apm.rum")},
        - 		"data_stream.namespace":           []any{string("default")},
        - 		"data_stream.type":                []any{string("traces")},
        - 		"destination.address":             []any{string("localhost")},
        - 		"destination.port":                []any{float64(8000)},
        - 		"event.outcome":                   []any{string("unknown")},
        - 		"http.response.decoded_body_size": []any{float64(676864)},
        - 		"http.response.encoded_body_size": []any{float64(676864)},
        - 		"http.response.transfer_size":     []any{float64(677175)},
        - 		"labels.testTagKey":               []any{string("testTagValue")},
        - 		"network.connection.type":         []any{string("5G")},
        - 		...
        - 	},
          	map[string]any{"@timestamp": []any{string("dynamic")}, "agent.name": []any{string("js-base")}, "agent.name.text": []any{string("js-base")}, "agent.version": []any{string("4.8.1")}, ...},
          	map[string]any{"@timestamp": []any{string("dynamic")}, "agent.name": []any{string("js-base")}, "agent.name.text": []any{string("js-base")}, "agent.version": []any{string("4.8.1")}, ...},
          }
        
        Test failed. Run `make check-approvals` to verify the diff.
        
    server.go:144: log file: /home/runner/work/apm-server/apm-server/systemtest/logs/TestRUMRoutingIntegration/apm-server
--- FAIL: TestRUMRoutingIntegration (3.99s)
```

--- a/systemtest/estest/search.go
+++ b/systemtest/estest/search.go
@@ -77,7 +77,15 @@ func ExpectMinDocs(t testing.TB, es *espoll.Client, min int, index string, query
 	return result
 }
 
-func ExpectSourcemapError(t testing.TB, es *espoll.Client, index string, retry func(), query interface{}, updated bool) espoll.SearchResult {
+// ExpectSourcemapError retries sending events via retry until the sourcemap
+// fetcher is available (i.e. no "fetcher unavailable" errors appear in the
+// indexed documents), then returns the search result.
+//
+// minDocs is the minimum number of documents expected in the index after a
+// successful retry. Callers must pass the exact count they expect so that the
+// underlying poll does not return early when Elasticsearch has only indexed a
+// partial batch (see: TotalHitsCondition race).
+func ExpectSourcemapError(t testing.TB, es *espoll.Client, index string, minDocs int, retry func(), query interface{}, updated bool) espoll.SearchResult {
 	t.Helper()
 
 	deadline := time.After(5 * time.Second)
@@ -99,7 +107,7 @@ func ExpectSourcemapError(t testing.TB, es *espoll.Client, index string, retry f
 
 			retry()
 
-			result := ExpectDocs(t, es, index, query)
+			result := ExpectMinDocs(t, es, minDocs, index, query)
 
 			if isFetcherAvailable(t, result) {
 				assertSourcemapUpdated(t, result, updated)

--- a/systemtest/rum_test.go
+++ b/systemtest/rum_test.go
@@ -220,7 +220,11 @@ func TestRUMRoutingIntegration(t *testing.T) {
 		require.NoError(t, err)
 		defer resp.Body.Close()
 	}
-	result := estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "traces-apm.rum*", retry, nil, false)
+	// 9 = 1 transaction + 8 spans produced by testdata/intake-v3/rum_events.ndjson.
+	// Passing the exact count prevents ExpectSourcemapError from returning early
+	// when Elasticsearch has only indexed a partial batch (TotalHitsCondition race).
+	const wantDocs = 9
+	result := estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "traces-apm.rum*", wantDocs, retry, nil, false)
 	approvaltest.ApproveFields(
 		t, t.Name(), result.Hits.Hits, "@timestamp", "timestamp.us",
 		"source.port", "source.ip", "client.ip",

--- a/systemtest/sourcemap_test.go
+++ b/systemtest/sourcemap_test.go
@@ -43,14 +43,14 @@ func TestRUMErrorSourcemapping(t *testing.T) {
 			retry := func() {
 				systemtest.SendRUMEventsPayload(t, serverURL, "../testdata/intake-v2/errors_rum.ndjson")
 			}
-		result := estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "logs-apm.error-*", 1, retry, nil, true)
-		approvaltest.ApproveFields(
-			t, t.Name(), result.Hits.Hits,
-			// RUM timestamps are set by the server based on the time the payload is received.
-			"@timestamp", "timestamp.us",
-			// RUM events have the source IP and port recorded, which are dynamic in the tests
-			"client.ip", "source.ip", "source.port",
-		)
+			result := estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "logs-apm.error-*", 1, retry, nil, true)
+			approvaltest.ApproveFields(
+				t, t.Name(), result.Hits.Hits,
+				// RUM timestamps are set by the server based on the time the payload is received.
+				"@timestamp", "timestamp.us",
+				// RUM events have the source IP and port recorded, which are dynamic in the tests
+				"client.ip", "source.ip", "source.port",
+			)
 		}
 
 		t.Run("standalone", func(t *testing.T) {
@@ -216,8 +216,8 @@ func TestSourcemapFetcher(t *testing.T) {
 			retry := func() {
 				systemtest.SendRUMEventsPayload(t, srv.URL, "../testdata/intake-v2/errors_rum.ndjson")
 			}
-		estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "logs-apm.error-*", 1, retry, nil, true)
-	})
+			estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "logs-apm.error-*", 1, retry, nil, true)
+		})
 	}
 }
 

--- a/systemtest/sourcemap_test.go
+++ b/systemtest/sourcemap_test.go
@@ -43,14 +43,14 @@ func TestRUMErrorSourcemapping(t *testing.T) {
 			retry := func() {
 				systemtest.SendRUMEventsPayload(t, serverURL, "../testdata/intake-v2/errors_rum.ndjson")
 			}
-			result := estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "logs-apm.error-*", retry, nil, true)
-			approvaltest.ApproveFields(
-				t, t.Name(), result.Hits.Hits,
-				// RUM timestamps are set by the server based on the time the payload is received.
-				"@timestamp", "timestamp.us",
-				// RUM events have the source IP and port recorded, which are dynamic in the tests
-				"client.ip", "source.ip", "source.port",
-			)
+		result := estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "logs-apm.error-*", 1, retry, nil, true)
+		approvaltest.ApproveFields(
+			t, t.Name(), result.Hits.Hits,
+			// RUM timestamps are set by the server based on the time the payload is received.
+			"@timestamp", "timestamp.us",
+			// RUM events have the source IP and port recorded, which are dynamic in the tests
+			"client.ip", "source.ip", "source.port",
+		)
 		}
 
 		t.Run("standalone", func(t *testing.T) {
@@ -91,7 +91,7 @@ func TestRUMSpanSourcemapping(t *testing.T) {
 	retry := func() {
 		systemtest.SendRUMEventsPayload(t, srv.URL, "../testdata/intake-v2/transactions_spans_rum_2.ndjson")
 	}
-	result := estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "traces-apm*", retry, espoll.TermQuery{
+	result := estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "traces-apm*", 1, retry, espoll.TermQuery{
 		Field: "processor.event",
 		Value: "span",
 	}, true)
@@ -124,7 +124,7 @@ func TestNoMatchingSourcemap(t *testing.T) {
 	retry := func() {
 		systemtest.SendRUMEventsPayload(t, srv.URL, "../testdata/intake-v2/transactions_spans_rum_2.ndjson")
 	}
-	result := estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "traces-apm*", retry, espoll.TermQuery{
+	result := estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "traces-apm*", 1, retry, espoll.TermQuery{
 		Field: "processor.event",
 		Value: "span",
 	}, false)
@@ -157,11 +157,11 @@ func TestSourcemapCaching(t *testing.T) {
 	retry := func() {
 		systemtest.SendRUMEventsPayload(t, srv.URL, "../testdata/intake-v2/errors_rum.ndjson")
 	}
-	estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "logs-apm.error-*", retry, nil, true)
+	estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "logs-apm.error-*", 1, retry, nil, true)
 
 	// Delete the source map and error, and try again.
 	systemtest.CleanupElasticsearch(t)
-	estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "logs-apm.error-*", retry, nil, true)
+	estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "logs-apm.error-*", 1, retry, nil, true)
 }
 
 func TestSourcemapFetcher(t *testing.T) {
@@ -216,8 +216,8 @@ func TestSourcemapFetcher(t *testing.T) {
 			retry := func() {
 				systemtest.SendRUMEventsPayload(t, srv.URL, "../testdata/intake-v2/errors_rum.ndjson")
 			}
-			estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "logs-apm.error-*", retry, nil, true)
-		})
+		estest.ExpectSourcemapError(t, systemtest.Elasticsearch, "logs-apm.error-*", 1, retry, nil, true)
+	})
 	}
 }
 

